File: apps/server/src/modules/vehicle-trip/vehicle-trip.service.ts
/**
 * VehicleTrip 服务
 * 包含车辆行程模块的业务逻辑
 */
import { Injectable, HttpException, HttpStatus } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import axios, { AxiosInstance } from 'axios';
import { VehicleTrip } from './vehicle-trip.entity';
import { SyncTripDto } from './dto/sync-trip.dto';
import { QueryTripDto } from './dto/query-trip.dto';
import { CommonConfigService } from '../common-config/common-config.service';

@Injectable()
export class VehicleTripService {
  constructor(
    @InjectRepository(VehicleTrip)
    private readonly vehicleTripRepository: Repository<VehicleTrip>,
    private readonly commonConfigService: CommonConfigService,
  ) {}

  /**
   * 获取 HTTP 客户端实例（动态读取配置中的 token）
   */
  private async getHttpClient(): Promise<AxiosInstance> {
    // 从配置表读取 token
    const config = await this.commonConfigService.findByKey('WanCheBaoToken');
    if (!config) {
      throw new HttpException('万车宝 Token 配置不存在', HttpStatus.INTERNAL_SERVER_ERROR);
    }
    if (!config.isEnabled) {
      throw new HttpException('万车宝 Token 配置未启用', HttpStatus.INTERNAL_SERVER_ERROR);
    }

    return axios.create({
      baseURL: 'https://online.wanchebao.com',
      headers: {
        'Host': 'online.wanchebao.com',
        'Accept': '*/*',
        'appName': '',
        'appVer': '3.0.5',
        'mobileModel': 'iPhone 13',
        'os': 'iOS',
        'mobileBrand': 'iPhone',
        'Accept-Language': 'zh-Hans-CN;q=1, en-CN;q=0.9',
        'osVer': '18.6.2',
        'token': config.configValue,
        'User-Agent': 'Advancer AD10/3.0.5 (iPhone; iOS 18.6.2; Scale/3.00)',
        'lang': 'CN',
        'Content-Type': 'application/json',
        'Cookie': 'JSESSIONID=9BD5378F7D8AD5AC59E6A18ED16F7112',
      },
    });
  }

  /**
   * 同步指定月份的车辆行程数据
   */
  async syncTripData(syncTripDto: SyncTripDto): Promise<{ success: number; failed: number; message: string }> {
    const { month } = syncTripDto;

    // 解析月份，计算开始和结束时间戳
    const { startTime, endTime } = this.parseMonthToTimestamp(month);

    try {
      // 调用外部API获取数据
      const httpClient = await this.getHttpClient();
      const response = await httpClient.get('/v2/driveRecord/section', {
        params: {
          vehicleId: 2032011,
          page: 1,
          limit: 1000,
          startTime,
          endTime,
          lang: 'CN',
        },
      });

      if (response.data.code !== 0) {
        throw new HttpException(
          `外部API返回错误: ${response.data.msg || '未知错误'}`,
          HttpStatus.BAD_GATEWAY,
        );
      }

      const tripDataList = response.data.data || [];
      let successCount = 0;
      let failedCount = 0;

      // 批量处理数据
      for (const item of tripDataList) {
        try {
          await this.saveTripData(item);
          successCount++;
        } catch (error) {
          console.error(`保存行程数据失败 (externalId: ${item.id}):`, error);
          failedCount++;
        }
      }

      return {
        success: successCount,
        failed: failedCount,
        message: `同步完成：成功 ${successCount} 条，失败 ${failedCount} 条`,
      };
    } catch (error) {
      if (error instanceof HttpException) {
        throw error;
      }
      throw new HttpException(
        `同步数据失败: ${error.message}`,
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }

  /**
   * 解析月份字符串，返回该月的开始和结束时间戳（毫秒）
   */
  private parseMonthToTimestamp(month: string): { startTime: number; endTime: number } {
    const [year, monthNum] = month.split('-').map(Number);
    const startDate = new Date(year, monthNum - 1, 1, 0, 0, 0, 0);
    const endDate = new Date(year, monthNum, 0, 23, 59, 59, 999);

    return {
      startTime: startDate.getTime(),
      endTime: endDate.getTime(),
    };
  }

  /**
   * 保存单条行程数据
   */
  private async saveTripData(item: any): Promise<VehicleTrip> {
    // 解析 raw 字段（JSON字符串）
    let rawData: any = {};
    try {
      rawData = typeof item.raw === 'string' ? JSON.parse(item.raw) : item.raw || {};
    } catch (error) {
      console.warn(`解析 raw 字段失败:`, error);
    }

    // 将外部ID（字符串）转换为数字
    const externalId = typeof item.id === 'string' ? parseInt(item.id, 10) : item.id;

    // 检查是否已存在（根据 externalId）
    const existingTrip = await this.vehicleTripRepository.findOne({
      where: { externalId },
    });

    // 构建实体数据
    const tripData: Partial<VehicleTrip> = {
      externalId,
      vehicleId: item.vehicleId,
      modelId: item.modelId,
      model: item.model,
      brandId: item.brandId,
      brand: item.brand,
      seriesId: item.seriesId,
      series: item.series,
      deviceId: item.deviceId,
      unitId: item.unitId,
      consumption: Math.round(item.consumption || 0),
      mileage: Math.round(item.mileage || 0),
      velocity: Math.round(item.velocity || 0),
      maxSpeed: Math.round(item.maxSpeed || 0),
      sharpAcceleration: item.sharpAcceleration || 0,
      sharpDeceleration: item.sharpDeceleration || 0,
      sharpTurn: item.sharpTurn || 0,
      startTime: item.startTime,
      endTime: item.endTime,
      // Raw 扁平化字段
      raw_start_time: this.parseNumber(rawData.start_time),
      raw_end_time: this.parseNumber(rawData.end_time),
      raw_te_record_trip_avg_oil: this.parseNumber(rawData.te_record_trip_avg_oil),
      raw_te_record_trip_avg_rpm: this.parseNumber(rawData.te_record_trip_avg_rpm),
      raw_te_record_trip_avg_speed: this.parseNumber(rawData.te_record_trip_avg_speed),
      raw_te_record_trip_max_rpm: this.parseNumber(rawData.te_record_trip_max_rpm),
      raw_te_record_trip_max_speed: this.parseNumber(rawData.te_record_trip_max_speed),
      raw_te_record_trip_mileage: this.parseNumber(rawData.te_record_trip_mileage),
      raw_te_record_trip_no: this.parseNumber(rawData.te_record_trip_no),
      raw_te_record_trip_oil: this.parseNumber(rawData.te_record_trip_oil),
      raw_te_record_trip_run_time: this.parseNumber(rawData.te_record_trip_run_time),
      raw_te_record_trip_start_time: this.parseNumber(rawData.te_record_trip_start_time),
      raw_te_record_trip_type: this.parseNumber(rawData.te_record_trip_type),
      raw_te_record_trip_urgent_acc_cnt: this.parseNumber(rawData.te_record_trip_urgent_acc_cnt),
      raw_te_record_trip_urgent_dec_cnt: this.parseNumber(rawData.te_record_trip_urgent_dec_cnt),
      raw_te_record_trip_urgent_turn_cnt: this.parseNumber(rawData.te_record_trip_urgent_turn_cnt),
    };

    if (existingTrip) {
      // 更新已存在的数据
      Object.assign(existingTrip, tripData);
      return await this.vehicleTripRepository.save(existingTrip);
    } else {
      // 创建新数据
      const newTrip = this.vehicleTripRepository.create(tripData);
      return await this.vehicleTripRepository.save(newTrip);
    }
  }

  /**
   * 解析数字（支持字符串和数字类型）
   */
  private parseNumber(value: any): number {
    if (value === null || value === undefined || value === '') {
      return 0;
    }
    const num = typeof value === 'string' ? parseInt(value, 10) : Number(value);
    return isNaN(num) ? 0 : num;
  }

  /**
   * 查询车辆行程数据列表（支持分页和时间范围筛选）
   */
  async findAll(queryTripDto: QueryTripDto): Promise<{
    data: VehicleTrip[];
    pagination: {
      page: number;
      limit: number;
      total: number;
      totalPages: number;
    };
  }> {
    const {
      page = 1,
      limit = 10,
      startTime,
      endTime,
    } = queryTripDto;

    // 构建查询条件
    const queryBuilder = this.vehicleTripRepository.createQueryBuilder('trip');

    // 时间范围筛选
    if (startTime) {
      queryBuilder.andWhere('trip.startTime >= :startTime', { startTime });
    }
    if (endTime) {
      queryBuilder.andWhere('trip.startTime <= :endTime', { endTime });
    }

    // 排序：按开始时间倒序
    queryBuilder.orderBy('trip.startTime', 'DESC');

    // 分页
    const skip = (page - 1) * limit;
    queryBuilder.skip(skip).take(limit);

    // 并行执行查询和计数
    const [data, total] = await Promise.all([
      queryBuilder.getMany(),
      queryBuilder.getCount(),
    ]);

    // 计算总页数
    const totalPages = Math.ceil(total / limit);

    return {
      data,
      pagination: {
        page,
        limit,
        total,
        totalPages,
      },
    };
  }
}

File: apps/server-go/internal/service/vehicle_trip_service.go
package service

/**
 * 车辆行程服务
 * 封装车辆行程同步和查询业务逻辑
 */

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"math"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"

	"github.com/zhouxiaomi/zxm_utils_server/apps/server-go/internal/dto"
	"github.com/zhouxiaomi/zxm_utils_server/apps/server-go/internal/model"
	"github.com/zhouxiaomi/zxm_utils_server/apps/server-go/internal/repository"
)

type IVehicleTripService interface {
	SyncTripData(ctx context.Context, req dto.SyncTripRequest) (int, int, error)
	FindAll(ctx context.Context, req dto.QueryTripRequest) ([]model.VehicleTrip, dto.Pagination, error)
}

type vehicleTripService struct {
	trips  repository.IVehicleTripRepository
	config repository.ICommonConfigRepository
}

func NewVehicleTripService(
	trips repository.IVehicleTripRepository,
	config repository.ICommonConfigRepository,
) IVehicleTripService {
	return &vehicleTripService{
		trips:  trips,
		config: config,
	}
}

func (s *vehicleTripService) SyncTripData(ctx context.Context, req dto.SyncTripRequest) (int, int, error) {
	client, headers, err := s.buildWanCheBaoClient(ctx)
	if err != nil {
		return 0, 0, err
	}

	month := req.Month
	if month == "" {
		return 0, 0, errors.New("时间不能为空")
	}
	startTime, endTime := parseMonthToTimestamp(month)

	u, err := url.Parse("https://online.wanchebao.com/v2/driveRecord/section")
	if err != nil {
		return 0, 0, err
	}
	q := u.Query()
	q.Set("vehicleId", "2032011")
	q.Set("page", "1")
	q.Set("limit", "1000")
	q.Set("startTime", strconv.FormatInt(startTime, 10))
	q.Set("endTime", strconv.FormatInt(endTime, 10))
	q.Set("lang", "CN")
	u.RawQuery = q.Encode()

	httpReq, err := http.NewRequestWithContext(ctx, http.MethodGet, u.String(), nil)
	if err != nil {
		return 0, 0, err
	}
	for k, v := range headers {
		httpReq.Header.Set(k, v)
	}

	resp, err := client.Do(httpReq)
	if err != nil {
		return 0, 0, err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return 0, 0, err
	}

	var parsed struct {
		Code int             `json:"code"`
		Msg  string          `json:"msg"`
		Data []map[string]any `json:"data"`
	}
	if err := json.Unmarshal(body, &parsed); err != nil {
		return 0, 0, err
	}
	if parsed.Code != 0 {
		return 0, 0, &ExternalAPIError{Msg: fmt.Sprintf("外部API返回错误: %s", parsed.Msg)}
	}

	success := 0
	failed := 0
	for _, item := range parsed.Data {
		if err := s.saveTripData(ctx, item); err != nil {
			failed++
		} else {
			success++
		}
	}
	return success, failed, nil
}

func (s *vehicleTripService) FindAll(ctx context.Context, req dto.QueryTripRequest) ([]model.VehicleTrip, dto.Pagination, error) {
	var startPtr, endPtr *int64
	if req.StartTime > 0 {
		v := req.StartTime
		startPtr = &v
	}
	if req.EndTime > 0 {
		v := req.EndTime
		endPtr = &v
	}

	list, total, err := s.trips.FindPaged(ctx, req.Page, req.Limit, startPtr, endPtr)
	if err != nil {
		return nil, dto.Pagination{}, err
	}
	if req.Page <= 0 {
		req.Page = 1
	}
	if req.Limit <= 0 {
		req.Limit = 10
	}
	totalPages := int((total + int64(req.Limit) - 1) / int64(req.Limit))
	return list, dto.Pagination{
		Page:       req.Page,
		Limit:      req.Limit,
		Total:      total,
		TotalPages: totalPages,
	}, nil
}

func (s *vehicleTripService) buildWanCheBaoClient(ctx context.Context) (*http.Client, map[string]string, error) {
	cfg, err := s.config.FindByKey(ctx, "WanCheBaoToken")
	if err != nil {
		return nil, nil, err
	}
	if cfg == nil {
		return nil, nil, errors.New("万车宝 Token 配置不存在")
	}
	if !cfg.IsEnabled {
		return nil, nil, errors.New("万车宝 Token 配置未启用")
	}
	token := ""
	if cfg.ConfigValue != nil {
		token = *cfg.ConfigValue
	}

	headers := map[string]string{
		"Host":            "online.wanchebao.com",
		"Accept":          "*/*",
		"appName":         "",
		"appVer":          "3.0.5",
		"mobileModel":     "iPhone 13",
		"os":              "iOS",
		"mobileBrand":     "iPhone",
		"Accept-Language": "zh-Hans-CN;q=1, en-CN;q=0.9",
		"osVer":           "18.6.2",
		"token":           token,
		"User-Agent":      "Advancer AD10/3.0.5 (iPhone; iOS 18.6.2; Scale/3.00)",
		"lang":            "CN",
		"Content-Type":    "application/json",
		"Cookie":          "JSESSIONID=9BD5378F7D8AD5AC59E6A18ED16F7112",
	}

	client := &http.Client{
		Timeout: 60 * time.Second,
	}
	return client, headers, nil
}

func (s *vehicleTripService) saveTripData(ctx context.Context, item map[string]any) error {
	rawField, _ := item["raw"]
	rawData := map[string]any{}
	switch v := rawField.(type) {
	case string:
		_ = json.Unmarshal([]byte(v), &rawData)
	case map[string]any:
		rawData = v
	}

	idStr := toString(item["id"])
	externalId, _ := strconv.ParseInt(idStr, 10, 64)

	existing, err := s.trips.FindByExternalID(ctx, externalId)
	if err != nil {
		return err
	}

	tripData := model.VehicleTrip{
		ExternalID: externalId,
		VehicleID:  int(parseInt(toString(item["vehicleId"]))),
		ModelID:    int(parseInt(toString(item["modelId"]))),
		Model:      toString(item["model"]),
		BrandID:    int(parseInt(toString(item["brandId"]))),
		Brand:      toString(item["brand"]),
		SeriesID:   int(parseInt(toString(item["seriesId"]))),
		Series:     toString(item["series"]),
		DeviceID:   toString(item["deviceId"]),
		UnitID:     int(parseInt(toString(item["unitId"]))),
		Consumption: int(roundFloat(toFloat(toString(item["consumption"])))),
		Mileage:     int(roundFloat(toFloat(toString(item["mileage"])))),
		Velocity:    int(roundFloat(toFloat(toString(item["velocity"])))),
		MaxSpeed:    int(roundFloat(toFloat(toString(item["maxSpeed"])))),
		SharpAcceleration: int(parseInt(toString(item["sharpAcceleration"]))),
		SharpDeceleration: int(parseInt(toString(item["sharpDeceleration"]))),
		SharpTurn:         int(parseInt(toString(item["sharpTurn"]))),
		StartTime:         int64(parseInt(toString(item["startTime"]))),
		EndTime:           int64(parseInt(toString(item["endTime"]))),
		RawStartTime:      int64(parseInt(toString(rawData["start_time"]))),
		RawEndTime:        int64(parseInt(toString(rawData["end_time"]))),
		RawAvgOil:         int(parseInt(toString(rawData["te_record_trip_avg_oil"]))),
		RawAvgRPM:         int(parseInt(toString(rawData["te_record_trip_avg_rpm"]))),
		RawAvgSpeed:       int(parseInt(toString(rawData["te_record_trip_avg_speed"]))),
		RawMaxRPM:         int(parseInt(toString(rawData["te_record_trip_max_rpm"]))),
		RawMaxSpeed:       int(parseInt(toString(rawData["te_record_trip_max_speed"]))),
		RawMileage:        int(parseInt(toString(rawData["te_record_trip_mileage"]))),
		RawNo:             int(parseInt(toString(rawData["te_record_trip_no"]))),
		RawOil:            int(parseInt(toString(rawData["te_record_trip_oil"]))),
		RawRunTime:        int(parseInt(toString(rawData["te_record_trip_run_time"]))),
		RawStartTime2:     int64(parseInt(toString(rawData["te_record_trip_start_time"]))),
		RawType:           int(parseInt(toString(rawData["te_record_trip_type"]))),
		RawUrgentAcc:      int(parseInt(toString(rawData["te_record_trip_urgent_acc_cnt"]))),
		RawUrgentDec:      int(parseInt(toString(rawData["te_record_trip_urgent_dec_cnt"]))),
		RawUrgentTurn:     int(parseInt(toString(rawData["te_record_trip_urgent_turn_cnt"]))),
	}

	if existing != nil {
		tripData.ID = existing.ID
		return s.trips.Update(ctx, &tripData)
	}
	return s.trips.Create(ctx, &tripData)
}

func parseMonthToTimestamp(month string) (int64, int64) {
	parts := strings.Split(month, "-")
	if len(parts) != 2 {
		return 0, 0
	}
	year, _ := strconv.Atoi(parts[0])
	monthNum, _ := strconv.Atoi(parts[1])
	// 使用本地时区计算月份起止时间，与 NestJS 的 new Date(year, month-1, ...) 行为保持一致
	startDate := time.Date(year, time.Month(monthNum), 1, 0, 0, 0, 0, time.Local)
	endDate := startDate.AddDate(0, 1, 0).Add(-time.Millisecond)
	return startDate.UnixMilli(), endDate.UnixMilli()
}

func toFloat(v string) float64 {
	if v == "" {
		return 0
	}
	f, _ := strconv.ParseFloat(v, 64)
	return f
}

func roundFloat(f float64) float64 {
	if f >= 0 {
		return math.Floor(f + 0.5)
	}
	return math.Ceil(f - 0.5)
}

